From 52050358e171be928a724d117a69c253c91e83ab Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Tue, 6 Dec 2022 14:36:50 +0100
Subject: [PATCH 166/529] usb: misc: onboard_usb_hub: keep power when
 usb_wakeup_enabled_descendants

There may be devices enabled for wakeup among hub descendants. So check
both controller and no descendants are wakeup-enabled to power off the hub.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I051ee94d7f9f4515aa16967184e4ac0b5be70137
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/279505
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/usb/misc/onboard_usb_hub.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

--- a/drivers/usb/misc/onboard_usb_hub.c
+++ b/drivers/usb/misc/onboard_usb_hub.c
@@ -98,13 +98,12 @@ static int __maybe_unused onboard_hub_su
 	mutex_lock(&hub->lock);
 
 	list_for_each_entry(node, &hub->udev_list, list) {
-		if (!device_may_wakeup(node->udev->bus->controller))
+		if (!device_may_wakeup(node->udev->bus->controller) &&
+		    !usb_wakeup_enabled_descendants(node->udev))
 			continue;
 
-		if (usb_wakeup_enabled_descendants(node->udev)) {
-			power_off = false;
-			break;
-		}
+		power_off = false;
+		break;
 	}
 
 	mutex_unlock(&hub->lock);
